Text/URI/Types.hs

```diff
@@ -117,6 +124,70 @@ makeAbsolute scheme URI {..} = URI
 isPathAbsolute :: URI -> Bool
 isPathAbsolute = either id (const True) . uriAuthority
 
+-- | 'relativeTo ref base' makes a new 'URI' from 'ref' relative to absolute URI 'base'.
+--
+-- See also: <https://tools.ietf.org/html/rfc3986#section-5.2>
+
+relativeTo :: URI -> URI -> URI
+relativeTo r base
+    | isJust (uriScheme r) =
+        r { uriPath      = removeDotSegments (uriPath r) }
+    | isRight (uriAuthority r) =
+        r { uriScheme    = uriScheme base
+          , uriPath      = removeDotSegments (uriPath r)
+          }
+    | isNothing (uriPath r) =
+        r { uriScheme    = uriScheme base
+          , uriAuthority = uriAuthority base
+          , uriPath      = uriPath base
+          , uriQuery     = if not (null (uriQuery r))
+                             then uriQuery r
+                             else uriQuery base
+          }
+    | isPathAbsolute r =
+        r { uriScheme    = uriScheme base
+          , uriAuthority = uriAuthority base
+          , uriPath      = removeDotSegments (uriPath r)
+          }
+    | otherwise =
+        r { uriScheme    = uriScheme base
+          , uriAuthority = uriAuthority base
+          , uriPath      = removeDotSegments (mergePaths base r)
+          }
+
+mergePaths :: URI -> URI -> Maybe (Bool, NonEmpty (RText 'PathPiece))
```

Comment:
> The type of `mergePaths` seems to be inadequate, since the output of such a function, as per section 5.2.3 of RFC 3986 does not amount only to path component as represented in the library but also to whether the resulting path is absolute.

I agree that according to the RFC `mergePaths` may make a relative URI absolute. To me, this function however is private (I had it in the `where` clause of `relativeTo` earlier). And that's why I chose to have it handled a level higher up (in `relativeTo`). Of course, it's your call if you do it here or not.

> Also what happens when base URI is not absolute? It seems like this should not be allowed...

The RFC indeed specifies that the scheme component of the base URI must be specified. So, no relative base URI's are allowed. I wonder if it would be nice to support them anyway? It could be useful to do:

```haskell
URI.relativeTo <$> URI.mkURI "baz#target" <*> URI.mkURI "/foo/bar"
```
and get "/foo/baz#target" back. I don't see any reason not to support it. You're right that we need to add test cases for this though.